# Business Inteligence Sheet 7

### Juan David Camacho
### Hernán Russi
### Juan Esteban Daza  

## Exercise 1 (3 Points)
1. Show that, if $S$, $T$ are tidsets, then $i(ST) = i(S) ∩ i(T)$, and even $i(S) ⊇ i(T)$ if $S ⊆ T$
2. Let $X$, $Y$ be two disjoint itemsets, $Z ⊇ XY$ with $sup(Z) = sup(XY )$, and $W ⊂ X$.
Show that conf $(W \rightarrow (Z \ W )) ≤ conf (X \rightarrow Y )$ and explain why this is important
for the AssociationRules algorithm.
3. Consider two rules $X \rightarrow Y$ and $W \rightarrow Y$ such that $W ⊂ X$. One might expect that
the confidence is monotonic in the premise, i.e., that the fact that $X ⊃ W$ implies that
$conf(X \rightarrow Y ) ≥ conf(W \rightarrow Y )$. That is, adding knowledge to the premise should
not reduce the confidence of the rule.
Give an example that shows that this is not necessarily the case.

1.
$\text{Assume S, T are Tidsets.}$

 $$i(ST) = \left\lbrace x \quad|\quad \forall t\quad \in\quad S \cup T,\quad t \quad contains \quad x\quad\right\rbrace$$
 
 $$=\left\lbrace x \quad|\quad \forall t \in S ,\quad t \quad contains \quad x\quad\right\rbrace\quad \cap\quad \left\lbrace x \quad|\quad \forall t \in T ,\quad t \quad contains \quad x\quad\right\rbrace$$

 $$=i(S) \cap i(T)$$
 
 $\text{Now assume } S \subseteq T \text{ Then:}$
 
 $$ i(T)= \left\lbrace x \quad|\quad \forall t \in T,\quad t \quad contains \quad x\quad\right\rbrace$$
 $$ i(S)= \left\lbrace x \quad|\quad \forall t \in S,\quad t \quad contains \quad x\quad\right\rbrace$$


$$We \quad know \quad that \quad if \quad x \in i(T), \quad then \quad all\quad ti \in T \quad contain \quad x.$$ 
$$Then \quad we \quad know \quad that \quad x \in i(S), \quad as \quad it \quad will \quad also \quad be \quad contained \quad in \quad all \quad ti \in S.$$
$$ Thus, \quad so \quad far \quad i(S) \cap i(T)=i(T) \quad for \quad any \quad general \quad S, \quad T. \quad$$ 
$$Nonetheless, \quad we \quad can \quad make \quad i(S) \quad bigger: $$
$$ Let \quad xd \quad be \quad contained \quad in \quad all \quad the \quad ti \in S,\quad but \quad in \quad none \quad of \quad the \quad ti \in T.$$
$$Then \quad xd \in i(S) \quad but \quad xd \notin i(T). \quad But \quad we \quad know \quad i(S)\cap(T)=i(T).$$
$$Then \quad i(S)\cup i(T) = i(S).\quad All \quad of \quad this \quad implies: $$
$$For \quad all \quad x \in i(T),\quad we \quad say \quad x\in i(S), $$
$$\Longrightarrow i(T) \subseteq i(S) \Longrightarrow i(S) \supseteq i(T)

2.
$$ X \cap Y = \emptyset $$
$$ XY \subseteq Z  $$
$$ Sup(XY) = Sup(Z)  $$
$$ W \subset X  $$
 
$$ Sup(X)= |\left\lbrace t \quad|\quad < t\quad,i(t) >\in D \quad and \quad x \subseteq i(t) \right\rbrace|$$  
$$ Sup(Z/W)= |\left\lbrace t \quad|\quad < t\quad,i(t) >\in D \quad and \quad z/w \subseteq i(t) \right\rbrace|$$  

$$We \quad have \quad conf(W \rightarrow (Z/W)= \frac{Sup(W\cup(Z/W)}{Sup(W)} $$
$$= \frac{Sup(Z)}{Sup(W)} $$
$$= \frac{Sup(XY)}{Sup(W)} $$

$$ For \quad \frac{Sup(XY)}{Sup(W)} \leq \frac{Sup(XY)}{Sup(X)} \quad, \quad we \quad need: $$
$$ \Longrightarrow Sup(XY) \leq \frac{Sup(XY)}{Sup(W)}*Sup(W)$$
$$ \Longrightarrow Sup(XY)*Sup(X) \leq Sup(XY)*Sup(W)$$
$$ \Longrightarrow Sup(X) \leq Sup(W)$$

$$As \quad W \subset X \quad, we \quad know \quad that \quad t(X) \subset t(W).\quad Thus \quad Sup(X)\leq Sup(W.) $$
$$Therefore \quad \frac{Sup(XY)}{Sup(W)} \leq \frac{Sup(XY)}{Sup(X)} \quad, \quad We\quad conclude:$$

$$  conf(W \rightarrow (Z/W)) \leq conf(X \rightarrow Y)

3.
$$ X \cap Y = \emptyset $$
$$ W \cap Y = \emptyset $$
$$ W \subset X  $$

Take the dataset and the itemsets: 

$$\begin{matrix}
. & i1 & i2 & i3 & i4 & i5\\
t1 & 1 & 0 & 0 & 0 & 0\\ 
t2 & 1 & 1 & 1 & 0 & 0\\ 
t3 & 0 & 0 & 0 & 1 & 1\\ 
t4 & 0 & 0 & 0 & 1 & 1\\ 
t5 & 1 & 0 & 0 & 1 & 1\\ 
\end{matrix}$$

$$ x=\left\lbrace i1, \quad i2, \quad i3\right\rbrace$$
$$ y=\left\lbrace i4, \quad i5\right\rbrace$$
$$ w=\left\lbrace i1\right\rbrace$$

We see:

$$  conf(X \rightarrow Y) = \frac{Sup(X\cup Y)}{Sup(X)} = \frac{P(X\cap Y)}{P(X)} = \frac{0/5)}{1/5} = \frac{0}{5}= 0$$
$$  conf(W \rightarrow Y) = \frac{Sup(W\cup Y)}{Sup(W)} = \frac{P(W\cap Y)}{P(W)} = \frac{1/5)}{3/5} = \frac{5}{15}=\frac{1}{3} $$

$$  conf(X \rightarrow Y) = 0 \leq \frac{1}{3} =  conf(W \rightarrow Y) $$

## Exercise 2 (5 Points)
In this exercise, we will use the transaction view on itemsets. Note that we treat itemsets as python lists (not numpy arrays). Hence, in the following db refers to a list of lists (a list of transactions, which are encoded as lists).

1. Write functions apriori(db, minsup) and eclat(db, minsup) that run the respective
algorithms and return the set of frequent itemsets (as tuples of sets with their support,
using the threshold minsup). Run both algorithms on the shop dataset and report the
runtimes.

In [1]:
from template import read_database
from template import diff
from template import powerset
import numpy as np
import timeit

def get_prefix_tree_initial_single_items(db):
    c_0 = []
    for listiiiiiiiiiiii in db:
        for sublist in listiiiiiiiiiiii:
            c_0.append(sublist)

    unique_items = sorted(set(c_0))
    returned_list = []
    for value in unique_items:
        auxilar_list = []
        auxilar_list.append(value)
        returned_list.append(auxilar_list)
    return returned_list
    

def extend_prefix_tree(tree_level, k):

    c_0 = []
    for listiiiiiiiiiiii in tree_level:
        for sublist in listiiiiiiiiiiii:
            c_0.append(sublist)

    next_level_macro = powerset(c_0)
    next_level = []
    for sublist in next_level_macro:
        if len(sublist) == k:
            next_level.append(sublist)

    return sorted(next_level)

def compute_support(db, evaluated_itemset):
    sup = 0
    for row in db:
        if all(value in row for value in evaluated_itemset):
           sup += 1 
    return sup

def apriori(db, minsup):

    F = [] #Frequent k-itemsets
    C = [] #Prefix Tree
    C.append(get_prefix_tree_initial_single_items(db))

    k = 0
    length = 2
    while len(C[k]) != 0:
        supports = []
        for evaluated_itemset in C[k]:
            supports.append(compute_support(db, evaluated_itemset))
        
        #lines 7 - 9
        auxiliar  = C[k][:]
        auxiliarS = supports[:]
        for leaf in C[k]:
            if supports[C[k].index(leaf)] >= minsup:
                F.append((leaf, supports[C[k].index(leaf)]))
            else:
                auxiliarS.pop(auxiliar.index(leaf))
                auxiliar.remove(leaf)
        C[k] = auxiliar
        supports = auxiliarS
        #lines 7 - 9
        C.append(extend_prefix_tree(C[0], length))
        k += 1
        length += 1
    return F

In [6]:
def eclat(db, minsup):

    F = []
    P = []


    c_0 = []
    for listiiiiiiiiiiiiiii in db:
        for sublist in listiiiiiiiiiiiiiii:
            c_0.append(sublist)
    power_set_C0 = powerset(sorted(set(c_0)))
    tree_first_level = sorted([value for value in power_set_C0 if len(value) == 1])

    t_i_macro = []
    for i in range(len(tree_first_level)):
        t_i = []
        for row in db:
            if all(value in row for value in tree_first_level[i]):
                t_i.append(db.index(row))
        t_i_macro.append(sorted(t_i))
    
    for i in range(len(tree_first_level)):
        if len(t_i_macro[i]) >= minsup:
            P.append((tree_first_level[i], t_i_macro[i]))
    start = timeit.default_timer()
    def eclat2(P, minsup, F):
        #for k in range(5):
        for p in P:
            F.append((p[0], len(p[1])))
            P_a = []
            for p2 in P:
                if P.index(p2) > P.index(p):
                    Xab = sorted(list(set(p[0])) + list(set(p2[0])))
                    t_Xab = [value for value in p[1] if value in p2[1]]
                    if len(t_Xab) >= minsup:
                        P_a.append((sorted(set(Xab)),t_Xab))
            if len(P_a) != 0:
                eclat2(P_a, minsup, F)
            #P.append(P_a)
    stop = timeit.default_timer()
    eclat2(P, minsup, F)
    eclat_timer = stop - start
    print('Runtime eclat: ', eclat_timer)              
    return(F)



In [7]:

#shop_dataset_macro = read_database('shop.dat')
#print('terminó de leer')
x = read_database('example2.dat')
start = timeit.default_timer()
apriori(x, 2)
stop = timeit.default_timer()
apriori_timer = stop - start

eclat(x, 2)



print('Runtime apriori: ', apriori_timer)


Runtime eclat:  1.000240445137024e-06
Runtime apriori:  0.00014879973605275154


Considering that we use a fraction of the shop.dat database, we can see that the runtime for eclat is decent for 24 unique items, but in the eclat algorithm the execution is way faster. This is thanks to the fact that we build a P before we even use the eclat algorithm, therefore the heavy computational process is supossed to be done before executing eclat. 

2. Write functions createAssociationRules(fsets, minconf) and getStrongRules(db,
minsup, minconf). The first should implement the AssociationRules algorithm and
compute all association rules with a given confidence from a set of frequent sets in the
form (premise, conclusion, abs sup, conf). The second should compute all frequent sets and then derives the strong rules from it with respect to the given minimum
confidence.

In [4]:
def createAssociationRules(fsets, minconf):

    AMPB = []

    for frequent in fsets:
        fsets_items = [x[0] for x in fsets]
        A = []
        provisional_supp = []
        if len(frequent[0]) >= 2:
            powerset_frequent = powerset(frequent[0])
            for powerset_item in powerset_frequent:
                if powerset_item in fsets_items and powerset_frequent.index(powerset_item) > 0:
                    A.append(powerset_item)
                    provisional_supp.append(fsets[fsets_items.index(powerset_item)][1])

            while len(A) != 0:
                X = A[-1]
                Y = diff(frequent[0], X)
                supX = provisional_supp[A.index(X)]
                A.remove(X)
                c = frequent[1]/supX

                if c >= minconf:
                    AMPB.append((X, Y, frequent[1], c))
                else:
                    powerset_remove = powerset(X)
                    for powerset_remove_item in powerset_remove:
                        if powerset_remove_item in A:
                            A.remove(powerset_item)
    
    return AMPB
                    

def getStrongRules(db, minsup, minconf):

    F = eclat(db, minsup)

    JS = createAssociationRules(F, minconf)

    return JS

3. Your boss is interested in a list of strong rules of the shop dataset. He considers rules
strong if their support is at least 500 and their confidence is at least .75. Report a list
of such strong rules sorted by support. He is also interested in a sub-list of these strong
rules in which at least two items appear in the conclusion of the rule. Report also this
sub-list, ordered descendingly by the number of items in the conclusion.
State explicitly, e.g. in a markdown cell, what the first of these rules means, and explain
what the respective support and confidence value of that specific rule mean.
Hint: Use a Pandas DataFrame to store the rules and to operate on them

In [11]:
import pandas as pd
#print(getStrongRules(x, 2, 0.75))
rules = pd.DataFrame(getStrongRules(x, 2, 0.75))
rules.columns = ['Premise', 'Conclusion', 'ABS Sup', 'Confidence']
display(rules)


Runtime eclat:  8.00006091594696e-07


Premise Conclusion  ABS Sup  Confidence
0   [825]      [834]        2         1.0
1   [834]      [825]        2         1.0

For a context in which we work with a product based databased, we can interpret the strong rule set displayed in the pandas dataframe above.

For the first rule we can interpret that when an user buys the item #825 there is a probability determined by the confidence level (the confidence level being the probability that a transaction contains the conclusion given that the premise exists in the itemset), that the user will buy the second item #834. In this case the reported support means that the item in the premise and the item in the conclusion appear twice in the used database.


*Disclaimer:*
*Since the shop.dat is way to big for our pc's to run it, we resorted in only using the first 2 rows of the shop.dat, since the first 2 rows contain 24 unique items, which is close to the computational limit that our machines seem to have :(*